File: python/036.py
#!/usr/bin/env python3
"""
Project Euler Problem 36: Double-base palindromes

The decimal number, 585 = 1001001001_2 (binary), is palindromic in both bases.
Find the sum of all numbers, less than one million, which are palindromic in base 10 and base 2.
"""


def is_palindrome(s: str) -> bool:
    """Check if a string is palindromic."""
    return s == s[::-1]


def to_binary(n: int) -> str:
    """Convert integer to binary string."""
    return bin(n)[2:]


def main():
    LIMIT = 1_000_000
    total = 0
    
    for num in range(LIMIT):
        decimal_str = str(num)
        if not is_palindrome(decimal_str):
            continue
        
        binary_str = to_binary(num)
        if is_palindrome(binary_str):
            total += num
    
    print(total)


if __name__ == "__main__":
    main()

File: c/036.c
/*
 * Project Euler 036 - Double-base Palindromes
 * Find the sum of all numbers less than one million which are
 * palindromic in both base 10 and base 2.
 */
#include <stdio.h>
#include <stdbool.h>
#include <string.h>

bool is_palindrome_str(const char *s, int len) {
    for (int i = 0; i < len / 2; i++) {
        if (s[i] != s[len - 1 - i]) return false;
    }
    return true;
}

bool is_decimal_palindrome(int n) {
    char buf[16];
    int len = sprintf(buf, "%d", n);
    return is_palindrome_str(buf, len);
}

bool is_binary_palindrome(int n) {
    if (n == 0) return true;
    char buf[32];
    int len = 0;
    int tmp = n;
    while (tmp > 0) {
        buf[len++] = '0' + (tmp & 1);
        tmp >>= 1;
    }
    /* buf is reversed, but palindrome check is symmetric */
    return is_palindrome_str(buf, len);
}

int main(void) {
    const int LIMIT = 1000000;
    long long sum = 0;

    for (int num = 1; num < LIMIT; num++) {
        if (is_decimal_palindrome(num) && is_binary_palindrome(num)) {
            sum += num;
        }
    }

    printf("%lld\n", sum);
    return 0;
}
